docs/architecture/adr-009-evidence-module.md

```diff
@@ -0,0 +1,205 @@
+
+```go
+type Handler func(ctx sdk.Context, evidence Evidence) sdk.Error
+```
+
+### Submission
+
+Assuming the `Evidence` is valid, the corresponding slashing penalty is invoked
+for the `Evidence`'s `Type`. Keep in mind the slashing penalty for any `Type` can
+be configured through governance.
+
+```go
+type MsgSubmitInfraction struct {
+	Infraction
+}
+
+func handleMsgSubmitInfraction(ctx sdk.Context, keeper Keeper, msg MsgSubmitEvidence) sdk.Result {
+	if err := keeper.SubmitInfraction(ctx, msg.Infraction); err != nil {
+		return err.Result()
+	}
+    
+	// emit events...
+
+	return sdk.Result{ 
+		// ...
+	}
+}
+```
+
+The `x/evidence` module's keeper is responsible for matching the `Evidence` against
+the module's router. Upon success the validator is slashed and the infraction is
+persisted. In addition, the validator is jailed is the `Evidence` type is configured
+to do so. 
+
+```go
+func (k Keeper) SubmitInfraction(ctx sdk.Context, infraction Infraction) sdk.Error {
+	handler := keeper.router.GetRoute(infraction.Evidence.Route())
+	if err := handler(cacheCtx, infraction.Evidence); err != nil {
+		return ErrInvalidEvidence(keeper.codespace, err.Result().Log)
+	}
+	
+	keeper.stakingKeeper.Slash(
+		ctx,
+		infraction.ConsensusAddress, 
+		infraction.InfractionHeight, 
+		infraction.Power, 
+		keeper.GetSlashingPenalty(ctx, infraction.Evidence.Type()),
+	)
+	
+	if infraction.Evidence.Jailable() {
+		keeper.stakingKeeper.Jail(ctx, infraction.ConsensusAddress)
+	}
+
+	keeper.setInfraction(ctx, infraction)
+}
+```
+
+### Genesis
+
+We require the the `x/evidence` module's keeper to keep an internal persistent
+mapping between `Evidence` types and slashing penalties represented as `InfractionPenalty`.
+
+```go
+var slashingPenaltyPrefixKey = []byte{0x01}
+
+type InfractionPenalty struct {
+	EvidenceType    string
+	Penalty         sdk.Dec
+}
+
+func GetSlashingPenaltyKey(evidenceType string) []byte {
+	return append(slashingPenaltyPrefixKey, []byte(evidenceType)...)
+}
+
+func (k Keeper) GetSlashingPenalty(ctx sdk.Context, evidenceType string) sdk.Dec {
+	store := ctx.KVStore(k.storeKey)
+	
+	bz := store.Get(GetSlashingPenaltyKey(evidenceType))
+	if len(bz) == 0 {
+		return sdk.ZeroDec()
+	}
+	
+	var ip InfractionPenalty
+	k.cdc.MustUnmarshalBinaryLengthPrefixed(bz, &ip)
+	
+	return ip.Penalty
+}
+```
+
+Finally, we need to represent the genesis state of the `x/evidence` module. The
+module only needs a list of all submitted valid infractions and the infraction
+penalties.
+
+```go
+type GenesisState struct {
+	Infractions         []Infraction
+	InfractionPenalties []InfractionPenalty
+}
+```
+
+## Consequences
+
+### Positive
+
+- Allows the state machine to process equivocations submitted on-chain and penalize
+validators based on agreed upon slashing parameters
+- Does not solely rely on Tendermint to submit evidence
+
+### Negative
+
+- No easy way to introduce new evidence types through governance on a live chain
```

Comment:
But the penalties could easily be altered, right? This doesn't seem too problematic to me.

Reply:
Yes, through governance.